helix-core/src/main/java/org/apache/helix/task/JobDispatcher.java

```diff
@@ -87,9 +87,10 @@ public ResourceAssignment processJobStatusUpdateAndAssignment(String jobName,
     TaskState jobState = workflowCtx.getJobState(jobName);
     // The job is already in a final state (completed/failed).
     if (workflowState == TaskState.FAILED || workflowState == TaskState.COMPLETED
-        || jobState == TaskState.FAILED || jobState == TaskState.COMPLETED) {
+        || jobState == TaskState.FAILED || jobState == TaskState.COMPLETED
+        || jobState == TaskState.TIMED_OUT) {
       LOG.info(String.format(
-          "Workflow %s or job %s is already failed or completed, workflow state (%s), job state (%s), clean up job IS.",
+          "Workflow %s or job %s is already in final state, workflow state (%s), job state (%s), clean up job IS.",
```

Comment:
maybe it is better to change this comment to failed, complete and time_out. Because there are other terminal states as well.

Reply:
The thing is Workflow = timed_out is not checked here.